R/combinedscatterplot.R

```diff
@@ -1093,3 +1137,146 @@ reorderPanels <- function(scatter.groups, x.order) {
+            }
+            if (!is.null(output.y$warning)) {
+                warning(output.y$warning)
+            }
+            x.midpoint = output.x$value
+            y.midpoint = output.y$value
+        }
+    }
+    list(quadrants.show = quadrants.show, x.midpoint = x.midpoint,
+         y.midpoint = y.midpoint)
+}
+
+computeMidpointValue <- function(midpoint.type, midpoint.input, midpoint.value,
+                                 data.values, axis, bounds.min, bounds.max) {
+    if (!is.numeric(data.values)) {
+        return(list(value = NaN,
+                    warning = paste0("Quadrants cannot be shown as the ", axis, "-axis has non-numeric data.")))
+    }
+
+    invalid.warning <- paste0("Quadrants cannot be shown as the ", axis, " midpoint value is invalid.")
+
+    # Estimate the range chosen by Plotly
+    estimated.range <- estimateRange(data.values, bounds.min, bounds.max)
+
+    out.of.range.warning <- paste0("The ", axis, " midpoint line is not shown as it is outside the plot range.")
+
+    if (midpoint.type == "Fixed value") {
+        if (is.null(midpoint.value)) {
+            return(list(value = NaN,
+                        warning = invalid.warning))
+        }
+        if (is.character(midpoint.value)) {
+            midpoint.value <- charToNumeric(midpoint.value)
+        }
+        if (is.na(midpoint.value) || !is.numeric(midpoint.value)) {
+            return(list(value = NaN,
+                        warning = invalid.warning))
+        }
+        if (midpoint.value < estimated.range$min || midpoint.value > estimated.range$max)  {
+            return(list(value = midpoint.value,
+                        warning = out.of.range.warning))
+        }
+        return(list(value = midpoint.value))
+    }
+
+    if (midpoint.type == "Calculation") {
+        if (is.null(midpoint.input) || !is.numeric(midpoint.input) || length(midpoint.input) == 0) {
+            return(list(value = NaN, warning = invalid.warning))
+        }
+
+        # Remove attributes from midpoint.input
+        midpoint.input <- as.numeric(midpoint.input)
+
+        if (length(midpoint.input) > 1 ) {
+            val <- midpoint.input[1]
+            if (is.na(val)) {
+                return(list(value = NaN, warning = invalid.warning))
+            }
+            return(list(value = val,
+                        warning = paste0("The input for the ", axis,
+                                         " midpoint has multiple elements. The first element will be used.")))
+        }
+        if (is.na(midpoint.input)) {
+            return(list(value = NaN, warning = invalid.warning))
+        }
+        if (midpoint.input < estimated.range$min || midpoint.input > estimated.range$max) {
+            return(list(value = midpoint.input,
+                        warning = out.of.range.warning))
+        }
+        return(list(value = midpoint.input))
+    }
+
+    if (midpoint.type == "Average") {
+        return(list(value = mean(data.values)))
```

Comment:
should we set `na.rm = TRUE`?

Reply:
I've assume that there are no missing values (I remove them before passing the data into the function). I can add `na.rm = TRUE` to make it work in that case (and also to calls to min and max)